debian-bo-master/bo/source/net/ncpfs-0.23/man/nprint.1
.TH NPRINT 1 12/27/1995 nprint nprint
.SH NAME
nprint \- NetWare print client
.SH SYNOPSIS
.B nprint
[
.B -S
.I server
] [
.B -h
] [
.B -U
.I user name
] [
.B -P
.I password
 | 
.B -n
] [
.B -C
] [
.B -q
.I queue name
] [
.B -d
.I job description
] [
.B -p
.I path name
] [
.B -b
.I banner name
] [
.B -f
.I file name in banner
] [
.B -l
.I lines
] [
.B -r
.I rows
] [
.B -c
.I copies
] [
.B -t
.I tab size
] [
.B -T
] [
.B -N
] [
.B -F
.I form number
] 
.B file

.SH DESCRIPTION
With
.B nprint,
you can print files on print queues of a NetWare file server. 

There are a lot of options, so you should probably wrap some default
configurations into some shell scripts.

.B nprint
looks up the file
.I $HOME/.nwclient
to find a file server, a user name and possibly a password. See
nwclient(5) for more information. Please note that the access
permissions of .nwclient MUST be 600, for security reasons.

.SH OPTIONS
.B file
.RS 3
.B file
is the name of the file you want to print. If file is '-', or no
filename is given, standard input is used.
.RE

.B -h
.RS 3
.B -h
is used to print out a short help text.
.RE

.B -S
.I server
.RS 3
.B server
is the name of the server you want to use.
.RE

.B -U
.I user
.RS 3
.B user
is the user name to use for the print request at the server.
.RE

.B -P
.I password
.RS 3
.B password
is the password to use for the print request at the server. If neither
.B -n
nor
.B -P
are given, and the user has no open connection to the server, nprint
prompts for a password.
.RE

.B -n
.RS 3
.B -n
should be given if no password is required for the print request.
.RE

.B -C
.RS 3
By default, passwords are converted to uppercase before they are sent
to the server, because most servers require this. You can turn off
this conversion by
.B -C.
.RE

.B -q
.I queue name
.RS 3
.B queue name 
is the name of the print queue to use at the print server. At
present, you must specify it in upper case characters.
.RE

.B -d
.I job description
.RS 3
.B job description
is the string that appears in pconsole when you list the jobs for the
print queue. It is also printed somewhere on the top of the banner
page.
.RE

.B -p
.I pathname
.RS 3
.B pathname
is the string you want to be printed on the banner page in the field
'path'.
.RE

.B -b
.I bannername
.RS 3
.B bannername
is the string (up to 12 characters) you want to appear on the lower
part of the banner page.
.RE

.B -f
.I filename
.RS 3
.B filename
is the string (up to 12 characters) you want to appear on the upper
part of the banner page. Default: the name of the file that is
printed, or 'stdin'.
.RE

.B -l
.I lines
.RS 3
.B lines
is the number of lines to put on one page. Default: 66
.RE

.B -r
.I rows
.RS 3
.B rows
is the number of rows to put on one page. Default: 80
.RE

.B -c
.I copies
.RS 3
.B copies
tells the print server to the specified number of copies. Default: 1
.RE

.B -t
.I tabs
.RS 3
.B tabs
is the number of spaces to print for a Tab-Character. Default: 8
.RE

.B -T
.RS 3
tells the print server to expand Tab-Character and use 8 spaces
.RE

.B -N
.RS 3
tells the print server 
.B not
to use Form Feeds
.RE

.B -F
.I form number
.RS 3
.B form number
is the the number of the form to be put into the printer. If it's
different from the one currently in the printer, your job is only
printed if a printer operator has put in the correct form.
.RE

.SH SEE ALSO
.B nwclient(5), slist(1), pqlist(1), ncpmount(8), ncpumount(8)

.SH CREDITS
nprint was written by Volker Lendecke (lendecke@namu01.gwdg.de)


debian-bo-master/bo/source/net/ncpfs-0.23/util/nprint.c
/*
 *  nprint.c
 *
 *  Send data to a NetWare print queue.
 *
 *  Copyright (C) 1995 by Volker Lendecke
 *
 */

#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>
#include <stdlib.h>
#include <ctype.h>

#include "ncplib.h"

static char *progname;
static void
usage(void);
static void help(void);

void
main(int argc, char *argv[])
{
	struct ncp_conn *conn;

	char default_queue[] = "*";
	char *queue = default_queue;

	int opt;

	struct ncp_bindery_object q;
	struct queue_job j;
	struct print_job_record pj;

	int written, read_this_time;

	char buf[8192];

	char *file_name;
	int file;
	long err;

	progname = argv[0];

	memzero(j); memzero(pj); memzero(q);

	if (   (argc == 2)
	    && (strcmp(argv[1], "-h") == 0))
	{
		help();
		exit(0);
	}

	if ((conn = ncp_initialize(&argc, argv, 1, &err)) == NULL)
	{
		com_err(argv[0], err, "in ncp_initialize");
		exit(1);
	}

	/*
	 * Fill in default values for print job
	 */
	j.j.TargetServerID = 0xffffffff; /* any server */
	/* at once */
	memset(&(j.j.TargetExecTime), 0xff, sizeof(j.j.TargetExecTime));
	j.j.JobType = htons(0);
	strcpy(j.j.JobTextDescription, "No Description");

	pj.Version = 0;
	pj.TabSize = 8;
	pj.Copies = htons(1);
	pj.CtrlFlags = 0;
	pj.Lines = htons(66);
	pj.Rows = htons(80);
	strcpy(pj.FnameHeader, "stdin");

	while ((opt = getopt(argc, argv, "hq:d:p:b:f:l:r:c:t:F:TN"))!=EOF)
	{
		switch (opt) {
		case 'h':
			help();
			ncp_close(conn);
			exit(1);
		case 'p':
			/* Path */
			pj.CtrlFlags |= PRINT_BANNER;
			if (strlen(optarg) >= sizeof(pj.Path))
			{
				strncpy(pj.Path, optarg,
					sizeof(pj.Path));
			}
			else
			{
				strcpy(pj.Path, optarg);
			}
			break;
		case 'b':
			/* Banner Name */
			pj.CtrlFlags |= PRINT_BANNER;
			if (strlen(optarg) >= sizeof(pj.BannerName))
			{
				strncpy(pj.BannerName, optarg,
					sizeof(pj.BannerName));
			}
			else
			{
				strcpy(pj.BannerName, optarg);
			}
			break;
		case 'f':
			/* File Name in Banner */
			pj.CtrlFlags |= PRINT_BANNER;
			if (strlen(optarg) >= sizeof(pj.FnameBanner))
			{
				strncpy(pj.FnameBanner, optarg,
					sizeof(pj.FnameBanner));
			}
			else
			{
				strcpy(pj.FnameBanner, optarg);
			}
			break;
		case 'l':
			/* lines, default: 66 */
			if ((atoi(optarg) < 0) || (atoi(optarg) > 65535))
			{
				fprintf(stderr,
					"invalid line number: %s\n", optarg);
				break;
			}
			pj.Lines = htons(atoi(optarg));
			pj.CtrlFlags |= EXPAND_TABS;
			break;
		case 'r':
			/* rows, default: 80 */
			if ((atoi(optarg) < 0) || (atoi(optarg) > 65535))
			{
				fprintf(stderr,
					"invalid row number: %s\n", optarg);
				break;
			}
			pj.Rows = htons(atoi(optarg));
			pj.CtrlFlags |= EXPAND_TABS;
			break;
		case 'c':
			/* copies, default: 1 */
			if ((atoi(optarg) < 0) || (atoi(optarg) > 65000))
			{
				fprintf(stderr,
					"invalid copies: %s\n", optarg);
				break;
			}
			pj.Copies = htons(atoi(optarg));
			pj.CtrlFlags |= EXPAND_TABS;
			break;
		case 't':
			/* tab size, default: 8 */
			if ((atoi(optarg) < 0) || (atoi(optarg) > 255))
			{
				fprintf(stderr,
					"invalid tab size: %s\n", optarg);
				break;
			}
			pj.TabSize = atoi(optarg);
			pj.CtrlFlags |= EXPAND_TABS;
			break;
		case 'T':
			/* expand tabs, default tabsize: 8 */
			pj.CtrlFlags |= EXPAND_TABS;
			break;
		case 'N':
			/* no form feed */
			pj.CtrlFlags |= NO_FORM_FEED;
			break;
		case 'F':
			/* Form number, default: 0 */
			if ((atoi(optarg) < 0) || (atoi(optarg) > 255))
			{
				fprintf(stderr,
					"invalid form number: %s\n", optarg);
				break;
			}
			j.j.JobType = htons(atoi(optarg));
			break;
		case 'q':
			/* Queue name to print on, default: '*' */
			if (strlen(optarg) >= NCP_BINDERY_NAME_LEN)
			{
				printf("queue name too long: %s\n",
				       optarg);
				ncp_close(conn);
				exit(1);
			}
			queue = optarg;
			break;
		case 'd':
			/* Job Description */
			pj.CtrlFlags |= PRINT_BANNER;
			if (strlen(optarg) >= sizeof(j.j.JobTextDescription))
			{
				strncpy(j.j.JobTextDescription, optarg,
					sizeof(j.j.JobTextDescription));
			}
			else
			{
				strcpy(j.j.JobTextDescription, optarg);
			}
			break;
			
		default:
			usage();
			ncp_close(conn);
			exit(1);
		}
	}

	if (optind != argc-1)
	{
		usage();
		ncp_close(conn);
		exit(1);
	}

	file_name = argv[optind];

	if (strcmp(file_name, "-") == 0)
	{
		file = 0;	/* stdin */
	}
	else
	{
		file = open(file_name, O_RDONLY, 0);
		if (file < 0)
		{
			perror("could not open file");
			ncp_close(conn);
			exit(1);
		}

		if (strlen(file_name) >= sizeof(pj.FnameHeader))
		{
			strncpy(pj.FnameHeader, file_name,
				sizeof(pj.FnameHeader));
		}
		else
		{
			strcpy(pj.FnameHeader, file_name);
		}

		if (strlen(pj.FnameBanner) == 0)
		{
			if (strlen(file_name) >= sizeof(pj.FnameBanner))
			{
				strncpy(pj.FnameBanner, file_name,
					sizeof(pj.FnameBanner));
			}
			else
			{
				strcpy(pj.FnameBanner, file_name);
			}
		}
	}

	memcpy(j.j.ClientRecordArea, &pj, sizeof(pj));

	str_upper(queue);

	if (ncp_scan_bindery_object(conn, 0xffffffff, NCP_BINDERY_PQUEUE,
				    queue, &q) != 0)
	{
		printf("could not find queue %s\n", queue);
		ncp_close(conn);
		exit(1);
	}
	
	if (ncp_create_queue_job_and_file(conn, q.object_id, &j) != 0)
	{
		printf("create error\n");
		ncp_close(conn);
		exit(1);
	}

	written = 0;
	do
	{
	        read_this_time = read(file, buf, sizeof(buf));
		if (read_this_time < 0)
		{
			break;
		}

		if (ncp_write(conn, j.file_handle,
			      written, read_this_time, buf) < read_this_time)
		{
			break;
		}

		written += read_this_time;
	} while (read_this_time > 0);

	close(file);

	if (ncp_close_file_and_start_job(conn, q.object_id, &j) != 0) {
		printf("close error\n");
		ncp_close(conn);
		return;
	}

	ncp_close(conn);
	return;
}

static void
usage(void)
{
	fprintf(stderr, "usage: %s [options] file\n", progname);
}

static void
help(void)
{
        printf("\n");
        printf("usage: %s [options] file\n", progname);
        printf("\n"
	       "-S server      Server name to be used\n"
               "-U username    Username sent to server\n"
               "-P password    Use this password\n"
               "-n             Do not use any password\n"
               "-C             Don't convert password to uppercase\n"
               "-q queue name  Name of the printing queue to use\n"
               "-d job desc    Job description\n"
               "-p path name   Path name to appear on banner\n"
               "-b bannername  Banner name (up to 12 chars)\n"
               "-f filename    Filename to appear on banner\n"
               "-l lines       Number of lines per page\n"
               "-r rows        Number of rows per page\n"
               "-t tab         Number of spaces per tab\n"
               "-T             Print server tab expantion\n"
               "-N             Surpress print server form feeds\n"
               "-F form #      Form number to print on\n"
               "-h             print this help text\n"
               "\n");
}
